### ISSUE TITLE
Create market simulator

### ISSUE THREAD

**Issue by marcdotson** [andrewheiss/why-donors-donate#24]
Look at creating it in R vs. Excel.

**Comment by marcdotson** [comment]
Here are the basic components we need:

- Configuration (this is what we'll want to have manipulated using a dashboard).
- Results (explored as part of a dashboard).
- Compute (done in the background using the betas from the model).

@sophie-cee et. al are going to look at flexdashboards or Shiny to figure out the simplest, best option for building the dashboard. @marcdotson will provide an example Excel market simulator to work from.

**Comment by marcdotson** [comment]
@sophie-cee do you have one of the pricing simulators from your class with Jeff? That should be a better place to start. If not, Samuel has one, would you connect with him?

**Comment by sophie-cee** [comment]
Yes, I'll look into it.

**Comment by marcdotson** [comment]
@jeffgallini has access to the repo now as well! Confirmed that Samuel has the market simulator. Get familiar with it and start building out the first version for us.

**Comment by marcdotson** [comment]
@jeffgallini @sophie-cee this might prove helpful, provided by none other than @andrewheiss himself: https://www.andrewheiss.com/blog/2020/01/01/flexdashboard-dynamic-data/.

**Comment by andrewheiss** [comment]
Oh mean, yeah, flexdashboard is probably the way to go here. Shiny is too intimidating for me, but with flexdashboard you can do Shiny-ish things like change inputs, etc. without actually building a complete Shiny app.

**Comment by andrewheiss** [comment]
And with this project, the data doesn't need to live on Google Drive or whatever—when you publish the dashboard/Rmd from RStudio, it'll upload any necessary files, like data, etc.

**Comment by marcdotson** [comment]
@sophie-cee @jeffgallini `market-simulator` branch created. Let's do this thang!

**Comment by sophie-cee** [comment]
@marcdotson, @jeffgallini and I are looking at getting the draws from the intercept only model. 

Is `analysis > output > posterior draws > intercept_noncentered.rds` the right file? Not sure what `i` and `j` represent. Just need some direction on how to figure out what the attribute level means are.

**Comment by marcdotson** [comment]
@sophie-cee that's the right file. We went over `i` and `j` when you were plotting. Check those scripts/comments first.

**Comment by marcdotson** [comment]
From `03_analysis.Rmd` in the `plot-marginals` section, `i` is the levels and `j` is the covariates (where `j = 1` for the intercept-only model).

**Comment by jeffgallini** [comment]
@sophie-cee  Loaded in the intercept model and got the means for the data, next we need to:
1. drop missing attribute levels for the inputs
2. set up the matrix multiplication for the betas
3. set the inputs to work with the matrix multiplication

**Comment by jeffgallini** [comment]
Added the levels in the simulator. Still trying to figure out how to make an object that interacts with the drop downs. Will probably make a dataframe with all possible outcomes and filter down based on the selections

**Comment by jeffgallini** [comment]
Figured out how to make a reactive data frame. Next step is the matrix multiplication @sophie-cee

**Comment by jeffgallini** [comment]
I am not exactly sure what is causing the problem with the formatting. I have determined that it has something to do with how the dashboard is interactive with the web site layout. When I pull the dashboard outside the web site layout, it works just fine. Everything else is updated.

**Comment by marcdotson** [comment]
I'll investigate how we might "turn on" multiple covariates in the market simulator.

**Comment by marcdotson** [comment]
Here's a walkthrough of each tab in the current Excel draft of the market simulator:

- Results has the matrix of market shares for each organization/row for each persona/column (i.e., each column sums to one). I have a few charts here, but you might prefer to just copy and paste these values into their own CSV and import into R to produce the results, including plots.
- Organizations includes all of the configurations for the 24 organizations you requested. I’ve created names along the bottom. You can modify as you’d like. Just specify 1 for the level you want active (all zeros for the hold-out level for every attribute except the organization itself). They’ve all been set to Red Cross. If you want to include more than 24 organizations, let me know.
- Personas includes all the configurations for the 32 organizations you requested. I’ve created names along the bottom and highlighted the covariates you wanted to modify. For all the rest, I’ve tried to provide an average. You can modify as you like. I’ve added some additional information on the ranges for each covariate, though may still want to reference experiment.md with the specifics of the survey when you do so. If you want to include more than 32 personas, let me know.
- Betas includes the posterior means for the betas (well, thetas). If we ever need an interval for the market shares, we can use the post-warm-up posterior draws instead of the means. That’ll need to be done in R.
- Compute includes the details for all the required computations. It’s as if each persona is a theoretical person, so we generate their “individual-level” betas and use that produce expected utilities that loop around to create market shares in Results.

**Comment by marcdotson** [comment]
From @evanwride:

Here is the draft so far of the shiny/flexdashboard walkthrough. I got the shiny walkthrough to work, but for some reason the layout of the flexdashboard wasn't working. I got it to work in my own personal repository, but couldn't figure out the problem when running it through why-donors-donate.

Either way, the code is correct and the information is there to explain how to use both of these dashboards.

Shiny: https://github.com/andrewheiss/why-donors-donate/blob/market-simulator/analysis/shiny_walkthrough.Rmd

Flexdashboard: https://github.com/andrewheiss/why-donors-donate/blob/market-simulator/analysis/flexdashboard_walkthrough.Rmd

**Comment by marcdotson** [comment]
Keep adding to the above walkthroughs (knitted as `github_document`s) and use the notes above on the Excel simulator to finish up the market simulator:

- [x] Get the right betas loaded (see `output` > `posterior_draws` > `public_political_social_charity_demo.rds`).
- [x] Use/update the sidebar to be able to specify personas using all the covariates in `public_political_social_charity_demo`.
- [x] Add the intermediate step of predicting persona betas to be used in computing market share.

**Comment by marcdotson** [comment]
Thinking about this a little more linearly:

- Get the *betas* from the model.
- Specify the *personas*.
- Compute the *predicted betas* for each persona using the *betas* and specified *personas*.
- Specify the *organizations*.
- Compute the *expected utilities* for each *organization* and *persona* using the specified *organizations* and *predicted betas*.
- Normalize the *expected utilities* into *market shares*.

**Comment by z-wix** [comment]
@evanwride File is updated with the persona calculations. Layout got a little messed up, some of the things I see that need to be changed:

- RenderValue Boxes can be removed (formatting doesn't work on website)
- Headers that flexdashboard usually hides are visible
- Pages set up is not working, change to tabs {.tabset .tabset-fade} need to figure out if we can do tabs within tabs
- First Page has an overall Market share with all the personas on the graph, this will be a global chart at top of page with the tabs below for organizations and tabs under those for personas
- Sidebar can be taken out (best to comment it out so we can use those later in possible persona configurations)

I have been using https://rmarkdown.rstudio.com/flexdashboard/index.html as a resource for flexdashboard elements that might be useful to add to the walkthru. There are examples and you can look at the source code. 

@marcdotson there has been a problem with integrating some flexdashboard elements (such as the value boxes) to the simulator on the website, I don't know if perhaps you knew the reason.

**Comment by marcdotson** [comment]
@z-wix @evanwride one thing to add to our discussion from Friday -- let's include the "hold-out" level for each of the attributes but set the corresponding value to zero. So, for example, with Issue Area, the hold-out level (which doesn't have a parameter estimate since all other parameter estimates are evaluated relative to that level) is "Emergency Response." When that level is selected the corresponding value is zero since the resulting organization/market share should have the associated betas zeroed out.

**Comment by z-wix** [comment]
@marcdotson I went through and checked to make sure the persona inputs matched the persona levels so they would influence the reactive element. You can go over it to now. Let me know if there are any issues I can fix.

**Comment by marcdotson** [comment]
@z-wix have you pushed the changes and checked the results against the Excel market simulator?

**Comment by z-wix** [comment]
I am pushed the changes. I am going to check the results with the Excel market simulator later tonight or tomorrow morning.

**Comment by andrewheiss** [comment]
Cool cool cool. I'm going to pull all the changes and play with the results on Monday and start making revisions/additions to the paper itself

**Comment by marcdotson** [comment]
It'll be fun to play with. @z-wix has done a great job in translating what we've been working on into Shiny. There should be a couple of PRs coming your way before too long, @andrewheiss.

**Comment by marcdotson** [comment]
@z-wix I've send you the Excel version of the market simulator to compare against. If you can match the persona and organization configurations to that version by default, it'll be easy to confirm the simulator is working, at least initially, and then we can make changes to be sure it's matching.

Also, the biggest thing I saw was that the organization (i.e., Red Cross, Amnesty, etc.) isn't part of the organization configuration tabs. It must have dropped off somewhere. Plus the width of the dropdowns and some of the spelling, capitalization consistency we discussed in our last call.

**Comment by z-wix** [comment]
@marcdotson after doing a lot of testing I was able to get the data directly from the market simulator to work using the matrix multiplication, however, while trying to recreate it in the simulator's reactive elements something was going wrong with organization 1. This afternoon I will be working on it after classes and Ill give you an update tonight.

**Comment by marcdotson** [comment]
@z-wix it works!

Thanks for walking through the cleaning up I'd done. After correcting the default issue you identified, I found the two culprits:

- I had introduced an error for one of the personas by double counting `per2_org3_exp` in `per2_share_denom`.
- The trickier error was that the worship attendance variable was spelled "attendence" in the reactive table and "attendance" in the inputs.

Moral of the story: Always check the actual output piecewise. I used `renderTable` to print out each piece (`gammas`, predicted `betas`, the column from `per_choice`, etc.) until I found the issue.

The last thing we need to do is some more thorough testing. Change an element in the Shiny simulator, change the same element in the Excel simulator, and make sure the market shares still match. Do this for a *lot* of possible changes, for both organization and persona configurations. As long as everything still checks out, we should be able to make a PR for @andrewheiss tomorrow.
